### PR TITLE
xbmgmt2 flash --scan update

### PR DESCRIPTION
**Pull request by AShivangi**
- display bdf of the card
- display information about the platform installed in the system

```
Card [0000:04:00.0]:
        Card type:              u250
        Flash type:             SPI
        Flashable partition running on FPGA:
                xilinx_u250_xdma_201830_2,[ID=0x5d14fbe6],[SC=UNKNOWN]
        Flashable partitions installed in system:
                xilinx_u250_xdma_201830_2,[ID=0x5d027a10],[SC=4.2.0]
```

### CODE REVIEW

File: src/runtime_src/core/tools/xbmgmt2/flash/firmware_image.cpp [hunk 1]
```diff
@@ -398,82 +409,58 @@ std::vector<DSAInfo> firmwareImage::installedDSA;
...
+
+    for (std::string t : suffix) {
+
+        std::regex e("^" FORMATTED_FW_DIR "/([^/]+)/([^/]+)/([^/]+)/.+\\." + t);
+        std::cmatch cm;
+
+        for (boost::filesystem::recursive_directory_iterator iter(formatted_fw_dir, 
+            boost::filesystem::symlink_option::recurse), recursive_end; iter != recursive_end;) {
+            std::string name = iter->path().string();
+            std::regex_match(name.c_str(), cm, e);
```
Comment (stsoe) [inline src/runtime_src/core/tools/xbmgmt2/flash/firmware_image.cpp]: is c_str() really necessary?

File: src/runtime_src/core/tools/xbmgmt2/flash/firmware_image.cpp [hunk 2]
```diff
@@ -398,82 +409,58 @@ std::vector<DSAInfo> firmwareImage::installedDSA;
...
+    if (!installedDSA.empty())
+        return installedDSA;
+
+    // for 2RP
+    p = FORMATTED_FW_DIR;
+	if (!boost::filesystem::is_directory(p))
+        return installedDSA;
+
+    boost::filesystem::path formatted_fw_dir(FORMATTED_FW_DIR);
+    std::vector<std::string> suffix = { XSABIN_FILE_SUFFIX, DSABIN_FILE_SUFFIX};
```
Comment (stsoe) [inline src/runtime_src/core/tools/xbmgmt2/flash/firmware_image.cpp]: It is really necessary to construct a vector of strings here?
Would this maybe work?

> for (auto t : { XSABIN_FILE_SUFFIX, DSABIN_FILE_SUFFIX} ) {
> ...
> }

File: src/runtime_src/core/tools/xbmgmt2/flash/firmware_image.cpp [hunk 3]
```diff
@@ -104,22 +104,22 @@ void getVendorBoardFromDSAName(std::string& dsa, std::string& vendor, std::strin
 
 void parseDSAFilename(std::string filename, uint64_t& vendor, uint64_t& device, uint64_t& subsystem, uint64_t &ts)
```
[the remark this answers is not in the document]
Comment (AShivangi) [reply src/runtime_src/core/tools/xbmgmt2/flash/firmware_image.cpp]: I'll add comments explaining the digits. I talked to Stephen about regex and he thinks boost tokenizer is cleaner than regex

File: src/runtime_src/core/tools/xbmgmt2/flash/firmware_image.cpp [hunk 4]
```diff
@@ -398,82 +409,58 @@ std::vector<DSAInfo> firmwareImage::installedDSA;
...
+
+    // for 2RP
+    p = FORMATTED_FW_DIR;
+	if (!boost::filesystem::is_directory(p))
+        return installedDSA;
+
+    boost::filesystem::path formatted_fw_dir(FORMATTED_FW_DIR);
+    std::vector<std::string> suffix = { XSABIN_FILE_SUFFIX, DSABIN_FILE_SUFFIX};
+
+    for (std::string t : suffix) {
```
[the remark this answers is not in the document]
Comment (AShivangi) [reply src/runtime_src/core/tools/xbmgmt2/flash/firmware_image.cpp]: that gives an error so changing it to const std::string& instead

File: src/runtime_src/core/tools/xbmgmt2/flash/firmware_image.cpp (hunk 1 above)
Comment (AShivangi) [reply src/runtime_src/core/tools/xbmgmt2/flash/firmware_image.cpp]: yes

File: src/runtime_src/core/tools/xbmgmt2/flash/firmware_image.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (stsoe) [reply src/runtime_src/core/tools/xbmgmt2/flash/firmware_image.cpp]: Yes, you are indeed tokenizing the string.  And I take back the comment of 0,16 those are just conversion args to strtoull.

File: src/runtime_src/core/tools/xbmgmt2/flash/firmware_image.cpp (hunk 1 above)
Comment (stsoe) [reply src/runtime_src/core/tools/xbmgmt2/flash/firmware_image.cpp]: I don't think so.  I think it is std::smatch sm to stay with std::string.  Please grep code base for std::smatch it's used in different places.

Comment (AShivangi) [reply src/runtime_src/core/tools/xbmgmt2/flash/firmware_image.cpp]: Oh ok. I haven't really explored regex so I don't know much about it. I made the change you asked. I'll have to go through the code again when verifying 2RP so I'll take a closer look then :)

**Pull request merged**